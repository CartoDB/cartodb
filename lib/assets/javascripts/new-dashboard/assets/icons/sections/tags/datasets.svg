<svg width="14" height="16" xmlns="http://www.w3.org/2000/svg">
  <path d="M7 .5C3.645.5.25 1.53.25 3.5v9c0 1.97 3.395 3 6.75 3s6.75-1.03 6.75-3v-9c0-1.97-3.395-3-6.75-3zM7 2c3.41 0 5.25 1.065 5.25 1.5C12.25 3.935 10.41 5 7 5S1.75 3.935 1.75 3.5C1.75 3.065 3.59 2 7 2zm0 12c-3.41 0-5.25-1.065-5.25-1.5v-1.05c1.302.692 3.28 1.05 5.25 1.05 1.97 0 3.948-.359 5.25-1.05v1.05c0 .435-1.84 1.5-5.25 1.5zm0-3c-3.41 0-5.25-1.065-5.25-1.5V8.45C3.052 9.142 5.03 9.5 7 9.5c1.97 0 3.948-.359 5.25-1.05V9.5c0 .435-1.84 1.5-5.25 1.5zm0-3C3.59 8 1.75 6.935 1.75 6.5V5.45C3.052 6.141 5.03 6.5 7 6.5c1.97 0 3.948-.359 5.25-1.05V6.5C12.25 6.935 10.41 8 7 8z" fill="#2E3C43" fill-rule="nonzero"/>
</svg>
