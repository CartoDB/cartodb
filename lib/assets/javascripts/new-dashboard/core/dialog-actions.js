// Dialogs
import Dialog from 'new-dashboard/components/Backbone/Dialog.vue';
import CreateDialog from 'new-dashboard/components/Backbone/Dialogs/CreateDialog';
import ChangePrivacy from 'new-dashboard/components/Backbone/Dialogs/ChangePrivacy';
import DeleteDialog from 'new-dashboard/components/Backbone/Dialogs/DeleteDialog';
import DuplicateMap from 'new-dashboard/components/Backbone/Dialogs/DuplicateMap';
import LockVisualization from 'new-dashboard/components/Backbone/Dialogs/LockVisualization';
import MapMetadata from 'new-dashboard/components/Backbone/Dialogs/MapMetadata';
import DatasetMetadata from 'new-dashboard/components/Backbone/Dialogs/DatasetMetadata';
import Share from 'new-dashboard/components/Backbone/Dialogs/Share';

export function editMapMetadata (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <MapMetadata
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, MapMetadata },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function editDatasetMetadata (dataset, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <DatasetMetadata
          :dataset="dataset"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['dataset'],
      components: { Dialog, DatasetMetadata },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { dataset }
  );
}

export function createMap (selectedDatasets, backgroundPollingView, mamufasImportView) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <CreateDialog :selectedItems="selectedItems" :backgroundPollingView="backgroundPollingView" :mamufasImportView="mamufasImportView" @close="$emit('close')"/>
      </Dialog>`,
      props: ['selectedItems', 'backgroundPollingView', 'mamufasImportView'],
      components: { Dialog, CreateDialog }
    },
    {
      selectedItems: selectedDatasets,
      backgroundPollingView,
      mamufasImportView
    }
  );
}

export function changePrivacy (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <ChangePrivacy
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, ChangePrivacy },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function shareVisualization (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <Share
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, Share },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function duplicateVisualization (visualization) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DuplicateMap :visualization="visualization" @close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, DuplicateMap }
    },
    { visualization }
  );
}

export function changeLockState (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockVisualization
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, LockVisualization },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function changeVisualizationsLockState (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockVisualization
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, LockVisualization },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function deleteVisualization (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function deleteVisualizations (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}
