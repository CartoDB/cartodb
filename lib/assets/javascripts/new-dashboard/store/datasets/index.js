import CartoNode from 'carto-node';
import toObject from 'new-dashboard/utils/to-object';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/filters';

const client = new CartoNode.AuthenticatedClient();

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'updated_at',
  orderDirection: 'desc'
};

const datasets = {
  namespaced: true,
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 1
  },
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setDatasets (state, datasets) {
      state.list = toObject(datasets.visualizations, 'id');
      state.metadata = {
        total_entries: datasets.total_entries,
        total_likes: datasets.total_likes,
        total_shared: datasets.total_shared,
        total_locked: datasets.total_locked,
        total_user_entries: datasets.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setDatasetAttributes (state, {datasetId, datasetAttributes}) {
      Object.assign(state.list[datasetId], datasetAttributes);
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setOrder (state, orderOptions) {
      state.order = orderOptions.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOptions.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage) {
      state.resultsPerPage = resultsPerPage;
    },

    updateLike (state, {datasetId, liked}) {
      state.list[datasetId].liked = liked;
    },

    updateNumberLikes (state, {datasetId, likes}) {
      state.list[datasetId].likes = likes;
    }
  },
  actions: {
    fetchDatasets (context) {
      context.commit('setFetchingState');

      const params = {
        ...Filters[context.state.filterType],
        types: 'table',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage,
        with_dependent_visualizations: 10
      };

      client.getVisualization('',
        params,
        function (err, _, data) {
          if (err) {
            context.commit('setRequestError', err);
            return;
          }
          context.commit('setDatasets', data);
          context.commit('setPagination', context.state.page);
        });
    },

    filterDatasets (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetchDatasets');
    },

    orderDatasets (context, orderOptions) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOptions);
      context.dispatch('fetchDatasets');
    },

    like (context, dataset) {
      const currentLikeStatus = dataset.liked;
      context.commit('updateLike', { datasetId: dataset.id, liked: true });
      client.like(dataset.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { datasetId: dataset.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { datasetId: dataset.id, likes: data.likes });
        }
      );
    },

    deleteLike (context, dataset) {
      const currentLikeStatus = dataset.liked;
      context.commit('updateLike', { datasetId: dataset.id, liked: false });
      client.deleteLike(dataset.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { datasetId: dataset.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { datasetId: dataset.id, likes: data.likes });
        }
      );
    },

    updateDataset (context, datasetOptions) {
      context.commit('setDatasetAttributes', datasetOptions);
    },

    setResultsPerPage (context, perPage) {
      context.commit('setResultsPerPage', perPage);
    },

    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', { order: options.order, direction: options.order_direction });
      context.dispatch('fetchDatasets');
    },

    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', { order: DEFAULT_VALUES.order, direction: DEFAULT_VALUES.orderDirection });
    }
  }
};

export default datasets;
