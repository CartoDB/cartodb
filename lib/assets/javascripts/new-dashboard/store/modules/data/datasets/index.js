import Filters from 'new-dashboard/core/configuration/filters';

import * as VisualizationActions from '../../../actions/visualizations';
import * as VisualizationMutations from '../../../mutations/visualizations';

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'favorited,updated_at',
  orderDirection: 'desc,desc'
};

const datasets = {
  namespaced: true,
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 1
  },
  mutations: {
    setFetchingState: VisualizationMutations.setFetchingState,
    setFilterType: VisualizationMutations.setFilterType,
    setOrder: VisualizationMutations.setOrder(DEFAULT_VALUES),
    setPagination: VisualizationMutations.setPagination,
    setResultsPerPage: VisualizationMutations.setResultsPerPage,
    setRequestError: VisualizationMutations.setRequestError,
    setVisualizations: VisualizationMutations.setVisualizations,
    updateNumberLikes: VisualizationMutations.updateNumberLikes,
    updateVisualization: VisualizationMutations.updateVisualization
  },
  actions: {
    deleteLike: VisualizationActions.deleteLike,
    filter: VisualizationActions.filter,
    like: VisualizationActions.like,
    order: VisualizationActions.order,
    resetFilters: VisualizationActions.resetFilters(DEFAULT_VALUES),
    setResultsPerPage: VisualizationActions.setResultsPerPage,
    setURLOptions: VisualizationActions.setURLOptions,
    updateVisualization: VisualizationActions.updateVisualization,

    fetch (context) {
      const params = {
        ...Filters[context.state.filterType],
        types: 'table',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage,
        with_dependent_visualizations: 10
      };

      VisualizationActions.fetchVisualizations(context, params);
    }
  }
};

export default datasets;
