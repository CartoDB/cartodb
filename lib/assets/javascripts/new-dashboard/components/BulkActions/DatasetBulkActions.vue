<template>
  <BulkActions :actions="actions[actionMode]" v-on="getEventListeners()"></BulkActions>
</template>

<script>
import BulkActions from 'new-dashboard/components/BulkActions/BulkActions';
import * as DialogActions from 'new-dashboard/core/dialog-actions';
import * as Table from 'new-dashboard/core/table';
import * as Visualization from 'new-dashboard/core/visualization';

export default {
  name: 'DatasetBulkActions',
  inject: ['backboneViews'],
  components: {
    BulkActions
  },
  props: {
    areAllDatasetsSelected: {
      type: Boolean,
      default: false
    },
    selectedDatasets: {
      type: Array,
      required: true
    }
  },
  computed: {
    actions () {
      return {
        single: [
          { name: this.$t('BulkActions.datasets.selectAllDatasets'), event: 'selectAll' },
          { name: this.$t('BulkActions.datasets.createMap'), event: 'createMap' },
          { name: this.$t('BulkActions.datasets.changeDatasetPrivacy'), event: 'changePrivacy', shouldBeHidden: this.isAnyShared },
          { name: this.$t('BulkActions.datasets.duplicateDataset'), event: 'duplicateDataset' },
          { name: this.$t('BulkActions.datasets.lockDataset'), event: 'lockDataset', shouldBeHidden: this.isAnyShared },
          { name: this.$t('BulkActions.datasets.deleteDataset'), event: 'deleteDataset', isDestructive: true, shouldBeHidden: this.isAnyShared }
        ],
        multiple: [
          { name: this.$t('BulkActions.datasets.selectAllDatasets'), event: 'selectAll', shouldBeHidden: this.areAllDatasetsSelected },
          { name: this.$t('BulkActions.datasets.deselectAllDatasets'), event: 'deselectAll' },
          { name: this.$t('BulkActions.datasets.createMap'), event: 'createMap' },
          { name: this.$t('BulkActions.datasets.lockDatasets'), event: 'lockDatasets', shouldBeHidden: this.isAnyShared },
          { name: this.$t('BulkActions.datasets.deleteDatasets'), event: 'deleteDatasets', isDestructive: true, shouldBeHidden: this.isAnyShared }
        ],
        lock: [
          { name: this.$t('BulkActions.datasets.selectAllDatasets'), event: 'selectAll' },
          { name: this.$t('BulkActions.datasets.duplicateDataset'), event: 'duplicateDataset' },
          { name: this.$t('BulkActions.datasets.unlockDataset'), event: 'unlockDataset' }
        ],
        multipleLock: [
          { name: this.$t('BulkActions.datasets.selectAllDatasets'), event: 'selectAll', shouldBeHidden: this.areAllDatasetsSelected },
          { name: this.$t('BulkActions.datasets.deselectAllDatasets'), event: 'deselectAll' },
          { name: this.$t('BulkActions.datasets.unlockDatasets'), event: 'unlockDatasets' }
        ]
      };
    },
    actionMode () {
      const isAnyDatasetLocked = this.selectedDatasets.filter(dataset => dataset.locked);

      if (isAnyDatasetLocked.length) {
        return isAnyDatasetLocked.length > 1 ? 'multipleLock' : 'lock';
      }

      return this.selectedDatasets.length > 1 ? 'multiple' : 'single';
    },
    isAnyShared () {
      return this.selectedDatasets.some(dataset => Visualization.isShared(dataset, this.$cartoModels));
    }
  },
  methods: {
    getActionHandlers () {
      return {
        deselectAll: () => {
          this.deselectAll();
        },
        fetchList: () => {
          this.$store.dispatch('datasets/fetchDatasets');
        },
        updateVisualization: (model) => {
          this.$store.dispatch('datasets/updateDataset', { datasetId: model.get('id'), datasetAttributes: model.attributes });
        }
      };
    },
    getEventListeners () {
      const events = this.actions[this.actionMode].map(action => action.event);

      return events.reduce(
        (eventListeners, action) => {
          eventListeners[action] = this[action].bind(this);
          return eventListeners;
        }, {}
      );
    },
    showModal (componentDefinition, componentPropsData) {
      this.$modal.show(
        componentDefinition,
        componentPropsData,
        { width: '100%', height: '100%' }
      );
    },
    selectAll () {
      this.$emit('selectAll');
    },
    deselectAll () {
      this.$emit('deselectAll');
    },
    createMap () {
      DialogActions.createMap.apply(this, [
        this.selectedDatasets,
        this.backboneViews.backgroundPollingView.getBackgroundPollingView(),
        this.backboneViews.mamufasImportView.getView()
      ]);
    },
    changePrivacy () {
      DialogActions.changePrivacy.apply(this, [this.selectedDatasets[0], this.getActionHandlers()]);
    },
    duplicateDataset () {
      const selectedDataset = this.selectedDatasets[0];
      const bgPollingView = this.backboneViews.backgroundPollingView.getBackgroundPollingView();

      bgPollingView._addDataset({
        type: 'duplication',
        table_name: `${Table.getUnqualifiedName(selectedDataset.name)}_copy`,
        value: selectedDataset.name,
        create_vis: false
      });
      this.deselectAll();
    },
    unlockDataset () {
      DialogActions.changeLockState.apply(this, [
        this.selectedDatasets[0],
        'datasets',
        this.getActionHandlers()
      ]);
    },
    lockDataset () {
      DialogActions.changeLockState.apply(this, [
        this.selectedDatasets[0],
        'datasets',
        this.getActionHandlers()
      ]);
    },
    deleteDataset () {
      DialogActions.deleteVisualization.apply(this, [
        this.selectedDatasets[0],
        'datasets',
        this.getActionHandlers()
      ]);
    },
    unlockDatasets () {
      DialogActions.changeVisualizationsLockState.apply(this, [
        this.selectedDatasets,
        'datasets',
        this.getActionHandlers()
      ]);
    },
    lockDatasets () {
      DialogActions.changeVisualizationsLockState.apply(this, [
        this.selectedDatasets,
        'datasets',
        this.getActionHandlers()
      ]);
    },
    deleteDatasets () {
      DialogActions.deleteVisualizations.apply(this, [
        this.selectedDatasets,
        'datasets',
        this.getActionHandlers()
      ]);
    }
  }
};
</script>
