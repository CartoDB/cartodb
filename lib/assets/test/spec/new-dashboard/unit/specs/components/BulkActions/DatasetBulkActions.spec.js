import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import DatasetBulkActions from 'new-dashboard/components/BulkActions/DatasetBulkActions';
import datasets from '../../fixtures/datasets';
import datasetsShared from '../../fixtures/datasetsShared';
import datasetsLocked from '../../fixtures/datasetsLocked';
import * as DialogActions from '../__mocks__/dialog-actions';

const $t = key => key;

const localVue = createLocalVue();
localVue.use(Vuex);

describe('DatasetBulkActions.vue', () => {
  it('should render all bulk actions', () => {
    const datasetBulkAction = createDatasetBulkActionComponent({selectedDatasets: []});
    expect(datasetBulkAction).toMatchSnapshot();
  });

  it('should render all multiple actions for all selected when there are no locked or shared datasets "Deselect all", "Create map", "Lock" and "Delete" actions', () => {
    const datasetBulkAction = createDatasetBulkActionComponent({ allSelected: true });
    expect(datasetBulkAction).toMatchSnapshot();
  });

  it('should render all single actions: "Select all", "Create map", "Change privacy", "Duplicate", "Lock", "Delete"', () => {
    const selectedDatasets = datasets.visualizations.slice(0, 1);
    const datasetBulkAction = createDatasetBulkActionComponent({ selectedDatasets });
    expect(datasetBulkAction).toMatchSnapshot();
  });

  it('should render all multiple actions: "Select all", "Deselect all", "Create map", "Lock", "Delete"', () => {
    const datasetBulkAction = createDatasetBulkActionComponent({});
    expect(datasetBulkAction).toMatchSnapshot();
  });

  it('should render single locked actions: "Select all", "Duplicate" and "Unlock"', () => {
    const selectedDatasets = datasetsLocked.visualizations.slice(0, 1);
    const datasetBulkAction = createDatasetBulkActionComponent({ selectedDatasets });
    expect(datasetBulkAction).toMatchSnapshot();
  });

  it('should render multiple locked actions when not all datasets have been selected: "Select all", Deselect all" and "Unlock"', () => {
    const selectedDatasets = datasetsLocked.visualizations.slice(0, 2);
    const datasetBulkAction = createDatasetBulkActionComponent({ selectedDatasets });
    expect(datasetBulkAction).toMatchSnapshot();
  });

  it('should render multiple actions for a single shared dataset: "Select All", "Create Map", "Duplicate"', () => {
    const selectedDatasets = datasets.visualizations.slice(0, 1);
    const datasetBulkAction = createDatasetBulkActionComponent({ selectedDatasets, isAnyShared: true });
    expect(datasetBulkAction).toMatchSnapshot();
  });

  it('should render multiple actions for selection containing shared datasets: "Select All", "Create Map", "Deselect All"', () => {
    const datasetBulkAction = createDatasetBulkActionComponent({ isAnyShared: true });
    expect(datasetBulkAction).toMatchSnapshot();
  });

  describe('computed', () => {
    it('should return "multiple" action name when more than one dataset is selected', () => {
      const datasetBulkAction = createDatasetBulkActionComponent({});
      const result = datasetBulkAction.vm.actionMode;
      expect(result).toBe('multiple');
    });

    it('should return "single" action name when there is only a dataset selected', () => {
      const selectedDatasets = datasets.visualizations.slice(0, 1);
      const datasetBulkAction = createDatasetBulkActionComponent({ selectedDatasets });
      const result = datasetBulkAction.vm.actionMode;
      expect(result).toBe('single');
    });
  });

  describe('methods', () => {
    let datasetBulkAction;
    beforeEach(() => {
      datasetBulkAction = createDatasetBulkActionComponent({});
    });

    it('should return all events listened', () => {
      const events = {
        'deleteDatasets': {},
        'deselectAll': {},
        'lockDatasets': {},
        'selectAll': {}
      };
      expect(datasetBulkAction.vm.getEventListeners()).toMatchObject(events);
    });

    it('should call the function to show modal', () => {
      datasetBulkAction.vm.showModal();
      expect(datasetBulkAction.vm.$modal.show).toHaveBeenCalled();
    });

    it('should call the function selectAll', () => {
      datasetBulkAction.vm.selectAll();
      expect(datasetBulkAction.emitted('selectAll').length).toBe(1);
    });

    it('should call the function deselectAll', () => {
      datasetBulkAction.vm.deselectAll();
      expect(datasetBulkAction.emitted('deselectAll').length).toBe(1);
    });

    it('should open create map', () => {
      datasetBulkAction.vm.createMap();
      expect(DialogActions.createMap).toHaveBeenCalled();
    });

    it('should open change privacy modal', () => {
      datasetBulkAction.vm.changePrivacy();
      expect(DialogActions.changePrivacy).toHaveBeenCalled();
    });

    it('should open duplicate dataset modal', () => {
      const selectedDatasets = datasetsShared.visualizations.slice(0, 1);
      const datasetBulkAction = createDatasetBulkActionComponent({ selectedDatasets });
      const spy = spyOn(datasetBulkAction.vm, 'deselectAll');

      datasetBulkAction.vm.duplicateDataset();

      const bgPollingView = datasetBulkAction.vm.backboneViews.backgroundPollingView.getBackgroundPollingView();
      expect(bgPollingView._addDataset).toHaveBeenCalledWith(expect.objectContaining({
        value: selectedDatasets[0].table.name
      }));
      expect(spy).toHaveBeenCalled();
    });

    it('should open change visualizations lock state dialog via unlockDatasets funcion', () => {
      datasetBulkAction.vm.unlockDatasets();
      expect(DialogActions.changeVisualizationsLockState).toHaveBeenCalled();
    });

    it('should open change visualizations lock state dialog via lockDatasets funcion', () => {
      datasetBulkAction.vm.lockDatasets();
      expect(DialogActions.changeVisualizationsLockState).toHaveBeenCalled();
    });

    it('should open change lock state dialog via lockDataset function', () => {
      datasetBulkAction.vm.lockDataset();
      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should open change lock state dialog via unlockDataset function', () => {
      datasetBulkAction.vm.unlockDataset();
      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should open delete visualization dialog', () => {
      const spy = spyOn(datasetBulkAction.vm, 'deselectAll');
      datasetBulkAction.vm.deleteDataset();
      expect(DialogActions.deleteVisualization).toHaveBeenCalled();
      expect(spy).toHaveBeenCalled();
    });

    it('should open delete visualization dialog', () => {
      datasetBulkAction.vm.deleteDataset();
      expect(DialogActions.deleteVisualization).toHaveBeenCalled();
    });

    it('should open delete visualizations dialog', () => {
      datasetBulkAction.vm.deleteDatasets();
      expect(DialogActions.deleteVisualizations).toHaveBeenCalled();
    });
  });
});

function createDatasetBulkActionComponent ({ selectedDatasets = datasets.visualizations, allSelected = false, isAnyShared = false }) {
  const _addDataset = jest.fn();
  const $modal = {
    show: jest.fn()
  };
  const $store = new Vuex.Store({
    state: {
      user: {
        isOutOfDatasetsQuota: false
      }
    }
  });
  return mount(DatasetBulkActions, {
    propsData: {
      selectedDatasets,
      areAllDatasetsSelected: allSelected
    },
    mocks: {
      $modal,
      $t,
      $store,
      backboneViews: {
        backgroundPollingView: {
          getBackgroundPollingView () {
            return { _addDataset };
          }
        },
        mamufasImportView: {
          getView () {
            return {};
          }
        }
      }
    },
    computed: {
      isAnyShared: () => isAnyShared
    }
  });
}
