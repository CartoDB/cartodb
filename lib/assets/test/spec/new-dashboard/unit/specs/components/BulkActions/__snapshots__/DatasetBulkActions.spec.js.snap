// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetBulkActions.vue should render all bulk actions 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.selectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.createMap
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.changeDatasetPrivacy
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.duplicateDataset
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.lockDataset
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.datasets.deleteDataset
  </button></section>
`;

exports[`DatasetBulkActions.vue should render all multiple actions for all selected when there are no locked or shared datasets "Deselect all", "Create map", "Lock datasets" and "Delete datasets" actions 1`] = `
<section class="bulk-actions">
  <!----><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.deselectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.createMap
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.lockDatasets
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.datasets.deleteDatasets
  </button></section>
`;

exports[`DatasetBulkActions.vue should render all multiple actions: "Select all", "Deselect all", "Create map", "Lock datasets", "Delete datasets" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.selectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.deselectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.createMap
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.lockDatasets
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.datasets.deleteDatasets
  </button></section>
`;

exports[`DatasetBulkActions.vue should render all single actions: "Select all", "Create map", "Change privacy", "Duplicate dataset", "Lock dataset", "Delete dataset" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.selectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.createMap
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.changeDatasetPrivacy
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.duplicateDataset
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.lockDataset
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.datasets.deleteDataset
  </button></section>
`;

exports[`DatasetBulkActions.vue should render multiple actions for a single shared dataset: "Select All", "Create Map", "Duplicate Dataset" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.selectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.createMap
  </button>
  <!----><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.duplicateDataset
  </button>
  <!---->
  <!---->
</section>
`;

exports[`DatasetBulkActions.vue should render multiple actions for selection containing shared datasets: "Select All", "Create Map", "Deselect All" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.selectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.deselectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.createMap
  </button>
  <!---->
  <!---->
</section>
`;

exports[`DatasetBulkActions.vue should render multiple locked actions when not all datasets have been selected: "Select all", Deselect all" and "Unlock datasets" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.selectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.deselectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.unlockDatasets
  </button></section>
`;

exports[`DatasetBulkActions.vue should render single locked actions: "Select all", "Duplicate dataset" and "Unlock dataset" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.selectAllDatasets
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.duplicateDataset
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.datasets.unlockDataset
  </button></section>
`;
